File: threads2/Deposito.java
public class Deposito {
    private int items = 0;
    private int capacidade = 5;
    private boolean podeRetirar = false;
    private boolean podeColocar = true;
    
    public synchronized int retirar() {

        if (!podeRetirar) {
            System.out.println("[ESPERA] Consumidor esperando para retirar...\n");
            try {
                wait();
            } catch(InterruptedException e) {
                System.out.println("Interrompido");
            }
        }

        if (items > 0) {

            items--;
            System.out.println("[RETIRADA] Caixa retirada: Sobram "+items+" caixas");

            podeColocar = true;

        }

        if (items == 0) {
            podeRetirar = false;
        }

        System.out.println("[FINAL] Consumidor terminou de retirar. podeRetirar = "+podeRetirar+"; podeColocar = "+podeColocar+'\n');

        notify();
        return 0;

    }

    public synchronized int colocar () {

        if (!podeColocar) {
            System.out.println("[ESPERA] Produtor esperando para colocar...\n");
            try {
                wait();
            } catch(InterruptedException e) {
                System.out.println("Interrompido");
            }
        }

        if (items < capacidade) {

            items++;
            System.out.println("[ARMAZENADA] Caixa armazenada: Passaram a ser "+items+" caixas");

            podeRetirar = true;

        }

        if (items == capacidade) {
            podeColocar = false;
        }

        System.out.println("[FINAL] Produtor terminou de colocar. podeRetirar = "+podeRetirar+"; podeColocar = "+podeColocar+'\n');

        notify();
        return 0;

    }

    public static void main(String[] args) {

        Deposito dep = new Deposito();
        Produtor prod = new Produtor(dep, 2);
        Consumidor cons = new Consumidor(dep, 1);
        Produtor prod1 = new Produtor(dep, 1);
        Consumidor cons1 = new Consumidor(dep, 5);

        prod.start();
        cons.start();
        prod1.start();
        cons1.start();

    }

}

public class Produtor extends Thread implements Runnable {

    private Deposito dep;
    private int tempo;

    public Produtor(Deposito d, int t){
        this.dep = d;
        this.tempo = t;
    }

    public void run() {

        while (true) {

            try {
                Thread.sleep(tempo*1000);
            } catch(InterruptedException e) {
                System.out.println("Interrompido");
            }

            dep.colocar();

        }

    }

}

public class Consumidor extends Thread implements Runnable {

    private Deposito dep;
    private int tempo;

    public Consumidor(Deposito d, int t){
        this.dep = d;
        this.tempo = t;
    }

    public void run() {

        while (true) {

            try {
                Thread.sleep(tempo*1000);
            } catch(InterruptedException e) {
                System.out.println("Interrompido");
            }

            dep.retirar();

        }

    }

}

File: threads2/deposito.py
import threading
import time

class Deposito:
    itens = 0
    capacidade = 5
    podeRetirar = False
    podeColocar = True
    monitor = threading.Condition()

    def retirar(self):
        with self.monitor: #equivalente a "syncronized"

            if not self.podeRetirar:
                print('[ESPERA] Consumidor esperando para retirar...\n')
                self.monitor.wait()

            if (self.itens > 0):
                self.itens -= 1
                print('[RETIRADA] Caixa retirada: Sobram '+str(self.itens)+' caixas');
                self.podeColocar = True

            if (self.itens == 0):
                self.podeRetirar = False

            print('[FINAL] Consumidor terminou de retirar. podeRetirar = '+str(self.podeRetirar)+'; podeColocar = '+str(self.podeColocar)+'\n')

            self.monitor.notify()
            
    def colocar(self):
        with self.monitor: #equivalente a "syncronized"

            if not self.podeColocar:
                print('[ESPERA] Produtor esperando para colocar...\n')
                self.monitor.wait()

            if (self.itens < self.capacidade):
                self.itens += 1
                print('[ARMAZENADA] Caixa armazenada: Passaram a ser '+str(self.itens)+' caixas');
                self.podeRetirar = True

            if (self.itens == self.capacidade):
                self.podeColocar = False

            print('[FINAL] Produtor terminou de colocar. podeRetirar = '+str(self.podeRetirar)+'; podeColocar = '+str(self.podeColocar)+'\n')

            self.monitor.notify()

class Produtor(threading.Thread):

    def __init__(self, d, t):
        threading.Thread.__init__(self)
        self.dep = d
        self.tempo = t

    def run(self):
        while not parar:
            time.sleep(self.tempo)
            self.dep.colocar()
            
class Consumidor(threading.Thread):

    def __init__(self, d, t):
        threading.Thread.__init__(self)
        self.dep = d
        self.tempo = t

    def run(self):
        while not parar:
            time.sleep(self.tempo)
            self.dep.retirar()

dep = Deposito()
prod = Produtor(dep, 2)
cons = Consumidor(dep, 1)
# prod1 = Produtor(dep, 1)
# cons1 = Consumidor(dep, 5)

parar = False

prod.start()
cons.start()
# prod1.start()
# cons1.start()

try:
    while True:
        time.sleep(2)
except KeyboardInterrupt:
    parar = True